crates/ide-assists/src/handlers/inline_const_as_literal.rs

```diff
@@ -0,0 +1,695 @@
+use syntax::{ast, AstNode};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: inline_const_as_literal
+//
+// Evaluate and inline const variable as literal.
+//
+// ```
+// const STRING: &str = "Hello, World!";
+//
+// fn something() -> &'static str {
+//     STRING$0
+// }
+// ```
+// ->
+// ```
+// const STRING: &str = "Hello, World!";
+//
+// fn something() -> &'static str {
+//     "Hello, World!"
+// }
+// ```
+pub(crate) fn inline_const_as_literal(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    let variable = ctx.find_node_at_offset::<ast::PathExpr>()?;
+
+    if let hir::PathResolution::Def(hir::ModuleDef::Const(konst)) =
+        ctx.sema.resolve_path(&variable.path()?)?
+    {
+        let konst_ty = konst.ty(ctx.sema.db);
+
+        let fuel = 20;
+
+        // FIXME: Add support to handle type aliases for builtin scalar types.
+        //
+        // There is no way to have a const static reference to a type that contains a interior
+        // mutability cell.
+        validate_type_recursively(ctx, Some(&konst_ty), false, fuel)?;
+
+        let expr = konst.value(ctx.sema.db)?;
+
+        // FIXME: Tuples and other sequence types will be inlined even though
+        // they might contain inner block expressions.
+        // e.g `(1, { 2 + 3 })` will be inline as it is.
+        let value = eval_and_inline_recursively(ctx, &konst, &expr, fuel)?;
+
+        let id = AssistId("inline_const_as_literal", AssistKind::RefactorInline);
+
+        let label = format!("Inline as literal");
+        let target = variable.syntax().text_range();
+
+        return acc.add(id, label, target, |edit| {
+            edit.replace(variable.syntax().text_range(), value);
+        });
+    }
+    None
+}
+
+fn eval_and_inline_recursively(
+    ctx: &AssistContext<'_>,
+    konst: &hir::Const,
+    expr: &ast::Expr,
+    fuel: i32,
+) -> Option<String> {
+    match (fuel > 0, expr) {
+        (true, ast::Expr::BlockExpr(block)) if block.is_standalone() => {
+            eval_and_inline_recursively(ctx, konst, &block.tail_expr()?, fuel - 1)
+        }
+        (_, ast::Expr::Literal(lit)) => Some(lit.to_string()),
```

Comment:
Are these necessary? I think `render_eval` should handle these as well.

Reply:
They should not be relevant anymore after #14947, so I've removed the whole function in favour of using `render_eval`.